### PR TITLE
Wrapper for OAuth URLs

### PR DESCRIPTION
**Pull request by shwatang**
### - Link added should be a valid URL and accessible throughout the org.
   ### - If the branch name contains hotfix / revert by default the BVT workflow check will pass.

| Test Case Document URL                        |
|-----------------------------------------------|
| https://docs.google.com/spreadsheets/d/1Zg9jLtF5elZHVyECO8jERagpO2yINxKyfnYj5Ke6S2M/edit#gid=0 |

### CODE REVIEW

File: src/main/java/com/razorpay/OAuthTokenClient.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    public OauthToken refreshToken(JSONObject request) throws RazorpayException {
+        validateRefreshTokenRequest(request);
+        request.put(GRANT_TYPE, REFRESH_TOKEN);
+        return post(null, Constants.TOKEN, request, Constants.AUTH);
+    }
+
+    public OauthToken revokeToken(JSONObject request) throws RazorpayException {
+        validateRevokeTokenRequest(request);
+        return patch(null, Constants.REVOKE, request, Constants.AUTH);
```
Comment (shwatang) [inline src/main/java/com/razorpay/OAuthTokenClient.java:65]: This would be a POST request instead.

File: src/main/java/com/razorpay/OAuthTokenClient.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                new ValidationConfig(STATE, Collections.singletonList(ValidationType.NON_EMPTY_STRING))
+        );
+    }
+
+    private List<ValidationConfig> getValidationsForAccessTokenRequest() {
+        return Arrays.asList(
+                new ValidationConfig(CLIENT_ID, Collections.singletonList(ValidationType.ID)),
+                new ValidationConfig(CLIENT_SECRET, Collections.singletonList(ValidationType.NON_EMPTY_STRING)),
+                new ValidationConfig(REDIRECT_URI, Arrays.asList(ValidationType.NON_EMPTY_STRING, ValidationType.URL)),
+                new ValidationConfig(MODE, Collections.singletonList(ValidationType.NON_EMPTY_STRING))
```
Comment (KUNAL13) [inline src/main/java/com/razorpay/OAuthTokenClient.java:106]: mode is optional - https://razorpay.com/docs/partners/platform/onboard-businesses/integrate-oauth/integration-steps/#request-parameters

File: src/main/java/com/razorpay/OAuthTokenClient.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                + "&client_id=" + clientId
+                + "&redirect_uri=" + redirectUri
+                + scopesArray
+                + "&state=" + state;
+
+        return AuthorizeUrl;
+    }
+
+    public OauthToken getAccessToken(JSONObject request) throws RazorpayException {
+        validateAccessTokenRequest(request);
```
Comment (KUNAL13) [inline src/main/java/com/razorpay/OAuthTokenClient.java:53]: here the grant type will be authorization_code?

File: src/main/java/com/razorpay/Utils.java [hunk 4]
```diff
@@ -48,6 +52,20 @@ public static boolean verifySignature(String payload, String expectedSignature,
     return isEqual(actualSignature.getBytes(), expectedSignature.getBytes());
   }
 
+  public static String generateOnboardingSignature(JSONObject attributes, String secret) {
```
Comment (KUNAL13) [inline src/main/java/com/razorpay/Utils.java:55]: @bhavyay - this will also change according to new impl?

File: src/main/java/com/razorpay/ValidationType.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.razorpay;
+
+public enum ValidationType {
```
Comment (KUNAL13) [inline src/main/java/com/razorpay/ValidationType.java:3]: can't we use some library for this?

**Review by KUNAL13 — COMMENTED** [review]
@bhavyay : setAccessToken method is not implemented

File: src/main/java/com/razorpay/OAuthTokenClient.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        return Arrays.asList(
+                new ValidationConfig(CLIENT_ID, Collections.singletonList(ValidationType.ID)),
+                new ValidationConfig(REDIRECT_URI, Arrays.asList(ValidationType.NON_EMPTY_STRING, ValidationType.URL)),
+                new ValidationConfig(SCOPES, Collections.singletonList(ValidationType.NON_NULL)),
+                new ValidationConfig(STATE, Collections.singletonList(ValidationType.NON_EMPTY_STRING))
+        );
+    }
+
+    private List<ValidationConfig> getValidationsForAccessTokenRequest() {
+        return Arrays.asList(
```
[the remark this answers is not in the document]
Comment (bhavyay) [reply src/main/java/com/razorpay/OAuthTokenClient.java:102]: Code is not mandatory. Get access token can accept two grant types client_credentials and authorization_code. Code is not required when creating access token using client_credentials grant type.

File: src/main/java/com/razorpay/Utils.java (hunk 4 above)
Comment (KUNAL13) [reply src/main/java/com/razorpay/Utils.java:55]: also this has to be live secret

File: src/main/java/com/razorpay/OAuthTokenClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (KUNAL13) [reply src/main/java/com/razorpay/OAuthTokenClient.java:102]: then lets take grant_type and accordingly add validations

File: src/main/java/com/razorpay/OAuthTokenClient.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    private void validateRevokeTokenRequest(JSONObject request) throws RazorpayException {
+        payloadValidator.validate(request, getValidationsForRevokeTokenRequest());
+    }
+
+    private List<ValidationConfig> getValidationsForAuthRequestURL() {
+        return Arrays.asList(
+                new ValidationConfig(CLIENT_ID, Collections.singletonList(ValidationType.ID)),
+                new ValidationConfig(REDIRECT_URI, Arrays.asList(ValidationType.NON_EMPTY_STRING, ValidationType.URL)),
+                new ValidationConfig(SCOPES, Collections.singletonList(ValidationType.NON_NULL)),
+                new ValidationConfig(STATE, Collections.singletonList(ValidationType.NON_EMPTY_STRING))
```
[the remark this answers is not in the document]
Comment (bhavyay) [reply src/main/java/com/razorpay/OAuthTokenClient.java:97]: Yes, state is always required.

File: src/main/java/com/razorpay/OAuthTokenClient.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+        String clientId = request.getString("client_id");
+        String redirectUri = request.getString("redirect_uri");
+        String state = request.getString("state");
+        String[] scopes = jsonArrayToStringArray(request.getJSONArray("scopes"));
+
+        String scopesArray = scopes.length > 0 ?
+                "&scope[]=" + String.join("&scope[]=", scopes) : "";
+
+        String AuthorizeUrl = "https://"
```
[the remark this answers is not in the document]
Comment (bhavyay) [reply src/main/java/com/razorpay/OAuthTokenClient.java]: Updated.

File: src/main/java/com/razorpay/OAuthTokenClient.java (hunk 2 above)
Comment (bhavyay) [reply src/main/java/com/razorpay/OAuthTokenClient.java]: Updated.

File: src/main/java/com/razorpay/OAuthTokenClient.java (hunk 3 above)
Comment (bhavyay) [reply src/main/java/com/razorpay/OAuthTokenClient.java:59]: Grant type can be authorization_code or client_credentials

File: src/main/java/com/razorpay/OAuthTokenClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bhavyay) [reply src/main/java/com/razorpay/OAuthTokenClient.java:108]: If we add validations for value of grant_type, if there is any change in grant_types in BE, it needs to be changed here as well and merchants should update the SDK version.

File: src/main/java/com/razorpay/Utils.java (hunk 4 above)
Comment (bhavyay) [reply src/main/java/com/razorpay/Utils.java:55]: When test client id is passed, test secret should be passed and live secret needs to be passed for live client id

Comment (bhavyay) [reply src/main/java/com/razorpay/Utils.java]: Updated the new implementation

File: src/main/java/com/razorpay/ApiClient.java [hunk 9]
```diff
@@ -31,27 +31,45 @@ class ApiClient {
 
   private final int STATUS_MULTIPLE_CHOICE = 300;
 
+  ApiClient() { }
```
Comment (KUNAL13) [inline src/main/java/com/razorpay/ApiClient.java:34]: why is this added?

Comment (shwatang) [reply src/main/java/com/razorpay/ApiClient.java:34]: I think its added for Oauth public APIs where we dont need to set any Auth.

File: src/main/java/com/razorpay/OAuthTokenClient.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (shwatang) [reply src/main/java/com/razorpay/OAuthTokenClient.java:108]: Have added validation to check for params based on grant_type

**Review by KUNAL13 — APPROVED** [review]

**Review by KUNAL13 — APPROVED** [review]
